File: code/game_engine/othello.py
from os import system as sys
from platform import system as system
from datetime import datetime as date
import copy as cp

class othello_engine(object):
    """
    The class is the the othello engine
    If you want to initialize the engine with an existed engine, set the parent= GameEngineFrom
    If you want to record the replay, set record = True
    If you want to see the showing of gameboard, set show = True
    """
    def __init__(self, parent = None, show = False, record = False):
        self.pieces = ['o', 'x', ' '] #pieces color
        self.show = show
        self.record = record
        if self.record:
            with open("Othello_Record.txt", "a+") as f:
                time = date.now()
                f.write("\n" + "--- "+ time.date + "-" + time.month + "-" + time.day + " " + time.hour + ":" + time.minute + ":" + time.second+ " ---\n")
                f.close()
        if parent == None:
            self.gameboard = []  # gameboard
            self.currentplayer = 1 # 1 is for black, 0 is for white
            self.stones = [0, 0] #stone number for each player
            self.validmoves = [0, 0] # possible moves for each player
            self.validboard = [[], []] # the possible moves's board for each player
            self.end = False
            self.count = 0
            self.replay = ""
            self.restart()
            self.winner = None # 0 for black, 1 for white, -1 for draw
            self.last_flips = []
        else:
            self.gameboard = cp.deepcopy(parent.get_board())
            self.currentplayer = cp.deepcopy(parent.get_currentplayer(number = True))
            self.stones = cp.deepcopy(parent.get_stones(return_all = True))
            self.validmoves = cp.deepcopy(parent.get_validmoves(return_all = True))
            self.validboard = cp.deepcopy(parent.get_validboard(return_all = True))
            self.end = cp.deepcopy(parent.finished())
            self.count = cp.deepcopy(parent.get_count())
            self.replay = cp.deepcopy(parent.get_replay())
            self.winner = cp.deepcopy(parent.get_winner())

    def get_board(self):
        return self.gameboard

    def get_last_flips(self):
        return self.last_flips

    def get_stones(self, player = None, return_all = False):
        if return_all:
            return self.stones
        if player == None:
            player = self.currentplayer
        return self.stones[player]

    def get_validboard(self, player = None, return_all = False):
        if return_all:
            return self.validboard
        if player == None:
            player = self.currentplayer
        return self.validboard[player]

    def get_validmoves(self, player = None, return_all = False):
        if return_all:
            return self.validmoves
        if player == None:
            player = self.currentplayer
        return self.validmoves[player]

    def get_currentplayer(self, number = True):
        if number:
            return self.currentplayer
        else:
            return self.pieces[self.currentplayer]

    def get_count(self):
        return self.count

    def get_replay(self):
        return self.replay

    def get_winner(self):
        return self.winner

    def set_show(self, show):
        self.show = show

    def set_record(self, record):
        self.record = record


    def print_replay(self):
        print self.count, self.replay

    def finished(self):
        return self.end

    # def display(self, force = False, board = None):
    #     if board == None:
    #         board = self.validboard[self.currentplayer]
    #     if force or self.show:
    #         if system() == 'Windows':
    #             sys("cls")
    #         else:
    #             sys("clear")
    #         for i in xrange(8):
    #             print str(7-i) + " " + ' '.join(self.pieces[j] for j in board[7-i])
    #         print "  0 1 2 3 4 5 6 7"
    #         print
    #         raw_input()

    def __display(self, force = False, board = None):
        if board == None:
            board = self.gameboard
        if force or self.show:
            if system() == 'Windows':
                sys("cls")
            else:
                sys("clear")
            for i in xrange(8):
                print str(7-i),
                for j in xrange(8):
                    print self.pieces[board[j][7-i]],
                print
            print "  0 1 2 3 4 5 6 7"
            print

    def __record(self):
        if self.record:
            with open("Othello_Record.txt", "a+") as f:
                f.write(str(self.count) + ": " + self.replay + "\n")
                f.close()

    # prepare the new board for the new game
    def restart(self):
        self.count += 1;
        self.replay  = ""
        self.end = False
        self.currentplayer = 1
        if len(self.gameboard) == 0:
            for i in xrange(8):
                self.gameboard.append([-1]*8)
                self.validboard[0].append([-1]*8)
                self.validboard[1].append([-1]*8)
        else:
            for i in xrange(8):
                self.gameboard[i] =[-1]*8
                self.validboard[0][i] = [-1]*8
                self.validboard[1][i] = [-1]*8
        self.gameboard[3][3] = 1
        self.gameboard[3][4] = 0
        self.gameboard[4][3] = 0
        self.gameboard[4][4] = 1
        self.__validateAndCount()
        self.__display()


    # This method is to undo one move. You should pass in the pieces been flipped in last turn
    def undo_move(self, x, y, flips):
        if self.gameboard[x][y] == -1:
            print "Undo fail! No piece here!"
            return
        self.currentplayer = self.gameboard[x][y]
        self.gameboard[x][y] = -1
        for i, j in flips:
            self.gameboard[i][j] = 1 - self.gameboard[i][j]
        self.__validateAndCount()
        self.end = False

    # Change the self.pieces in certain directions between target point and end point.
    # The @current_piece is the current color the target point, or assumed color of it during idnetifying whether the point will a valid move.
    # The @change is a boolean. True if it is a actual move, False if it is an assumed move.
    def __change_inside(self, target_x, target_y, end_x, end_y, current_piece, change):
        i = target_x
        j = target_y
        listx = []
        listy = []
        num = 0
        found = False
        while not found and (i != end_x or j != end_y):
            if end_x - i != 0:
                i = i + (end_x - i)/abs(end_x - i)
            if end_y - j != 0:
                j = j + (end_y - j)/abs(end_y - j)
            if self.gameboard[i][j] == -1:
                return False
            elif self.gameboard[i][j] == current_piece:
                found = True
                break
            listx.append(i)
            listy.append(j)
            num = num + 1
        if num == 0 or not found:
            return False
        elif change == True:
            for n in xrange(num):
                self.gameboard[listx[n]][listy[n]] = current_piece
                self.last_flips.append((listx[n], listy[n]))
        return True

    # Expand from the target point in 8 directions.
    # The @current_piece is the current color the target point, or assumed color of it during idnetifying whether the point will a valid move.
    # The @change is a boolean. True if it is a actual move, False if it is an assumed move.
    def __expand(self, target_x, target_y, current_piece, change):
        enable = False
        up = min(target_x, target_y)
        down = min(7 - target_x, 7 - target_y)
        left = min(target_x, 7 - target_y)
        right = min(7 - target_x, target_y)

        if self.__change_inside(target_x, target_y, 0, target_y, current_piece, change):
            enable = True
        if self.__change_inside(target_x, target_y, 7, target_y, current_piece, change):
            enable = True
        if self.__change_inside(target_x, target_y, target_x, 0, current_piece, change):
            enable = True
        if self.__change_inside(target_x, target_y, target_x, 7, current_piece, change):
            enable = True
        if self.__change_inside(target_x, target_y, target_x-up, target_y-up, current_piece, change):
            enable = True
        if self.__change_inside(target_x, target_y, target_x-left, target_y+left, current_piece, change):
            enable = True
        if self.__change_inside(target_x, target_y, target_x+down, target_y+down, current_piece, change):
            enable = True
        if self.__change_inside(target_x, target_y, target_x+right, target_y-right, current_piece, change):
            enable = True

        if (enable == True):
            return True
        else:
            return False

    # mark the valid possible moves on the xboard and oboard
    def __validateAndCount(self):
        self.stones[0] = 0
        self.stones[1] = 0
        self.validmoves[0] = 0
        self.validmoves[1] = 0
        for i in xrange(8):
            for j in xrange(8):
                for k in xrange(2):
                    self.validboard[k][i][j] = -1
                    if self.gameboard[i][j] == -1:
                        if self.__expand(i, j, k, False):
                            self.validboard[k][i][j] = k
                            self.validmoves[k]+= 1
                    elif self.gameboard[i][j] == k:
                        self.stones[k] += 1

    def update(self, x, y):
        if self.end == True:
            return False
        if self.validboard[self.currentplayer][x][y] != self.currentplayer:
            self.__display(force = True)
            print "Invalid move!:", x, y
            return False
        self.gameboard[x][y] = self.currentplayer
        self.replay = self.replay + " " + str(self.currentplayer) + str(x) + str(y)
        #update the self.gameboard
        self.last_flips = []
        self.__expand(x, y, self.currentplayer, True)
        #update the self.validboard and self.validmoves
        self.__validateAndCount()
        if self.validmoves[1 - self.currentplayer] != 0:
            self.currentplayer = 1 - self.currentplayer
        if self.stones[0] == 0 or self.stones[1] == 0 or sum(self.stones) == 64 or (self.validmoves[0]== 0 and self.validmoves[1] == 0):
            self.end = True
            self.__record()
            self.__display()
            if self.stones[1] > self.stones[0]:
                self.winner = 1
                self.replay = "Black" + self.replay
                return "Black"
            elif self.stones[1] < self.stones[0]:
                self.winner = 0
                self.replay = "White" + self.replay
                return "White"
            else:
                self.winner = -1
                self.replay = "Draw" + self.replay
                return "Draw"
        self.__display()
        return True

    def process_file(self, filename):
        fout = open(filename+"_replay", "w")
        with open(filename, "r") as f:
            for line in f:
                self.restart()
                output = ""
                moves = line.split()
                for move in moves:
                    self.update(ord(move[1]) - ord('1'), ord(move[0]) - ord('A'))
                fout.write(self.replay+"\n")
            f.close()
        fout.close()

if  __name__ == '__main__':
    cmd = raw_input("Start a new Game (y/n)? : ")
    engine = othello_engine(show = True)
    while cmd == "y" or cmd == "Y":
        engine.restart()
        while not engine.finished():
            print "currentplayer is", engine.get_currentplayer(number=False)
            cmd = raw_input("input x, y: ")
            if len(cmd) > 2:
                if cmd[0].isdigit() and cmd[2].isdigit():
                    engine.update(ord(cmd[0]) - ord('0'), ord(cmd[2]) - ord('0'))
        if engine.get_winner() == 0:
            print "Winner is Black!"
        elif engine.get_winner() == 1:
            print "Winner is White!"
        else:
            print "Draw!"
        cmd = raw_input("Start a new Game (y/n)? : ")

File: Website/html_public/js/othello.js
/*
  prepare the new board for the new game
*/
function restart() {
	currentplayer = pieces[0];
	for (var i = 0; i < 8; i++) {
		for (var j = 0; j < 8; j++) {
			gameboard[i][j] = ".";
		}
	}

	record = "";
	update_old();

	gameboard[3][3] = pieces[1];
	gameboard[3][4] = pieces[0];
	gameboard[4][3] = pieces[0];
	gameboard[4][4] = pieces[1];

	validateAndCount();
	display();
	handleEvent();
	$("#downloadRecord").remove();
}

/*
  let the oldboard updated with the gameboard
*/
function update_old() {
	for (var i = 0; i < 8; i++) {
		for (var j = 0; j < 8; j++) {
			oldboard[i][j] = gameboard[i][j];
		}
	}
}

/*
  clean the old element
*/
function cleanboard() {
	for (var i = 0; i < 8; i++) {
		for (var j = 0; j < 8; j++) {
			for (var k = 0; k < 2; k++) {
				var chess_id = elementName[k] + i + j;
				$('#' + chess_id).remove();
			}
		}
	}
}


/*
  display all the pieces and let the valid possible moves to be seen when hovered on
*/
function display() {

	var top = $("#chessboard").position().top; // top posotion of the whole chess board
	var left = $("#chessboard").position().left; // left posotion of the whole chess board
	var edge = 40;
	var chess_size = 65;
	if ($( window ).width() < 768){
		edge = 40/600 * $( window ).width();
		chess_size = 65/600 * $( window ).width();
	}

	var board = $("#board");
	board.css("top", top + "px");
	board.css("left", left + "px");

	cleanboard();
	update_old();

	for (var i = 0; i < 8; i++) {
		for (var j = 0; j < 8; j++) {
			for (var k = 0; k < 2; k++) {
				if (gameboard[i][j] == pieces[k]) {

					var chess_id = elementName[k] + i + j;

					var xcoor = left + edge + j * chess_size;
					var ycoor = top + edge + i * chess_size;
					$("#chessboard").append('<div class="chess" id="' + chess_id + '"></div>');
					$("#" + chess_id).css({
						"top": ycoor + "px",
						"left": xcoor + "px"
					});
					$("#" + chess_id).append('<img src="' + photoScource[k] + '">');
				}

				// here we want to add the possible moves to be elements
				if (validboard[k][i][j] == currentplayer) {
					var chess_id = elementName[k] + i + j;
					var xcoor = left + edge + j * chess_size;
					var ycoor = top + edge + i * chess_size;
					$("#chessboard").append('<div class="chess hover" id="' + elementName[k] + i + j + '"></div>');
					$("#" + chess_id).css({
						"top": ycoor + "px",
						"left": xcoor + "px"
					});
					$("#" + chess_id).append('<img src="' + photoScource[k + 2] + '">');
				}
			}
		}
	}

	$("#blackstones").text(stones[0]);
	$("#whitestones").text(stones[1]);
}

function handleEvent() {

	for (var k = 0; k < 2; k++) {
		$(".chess.hover").children('img').hide();
		if (currentplayer == pieces[k]) {
			if ((currentplayer == pieces[0]) == user_is_black){
				$(".chess.hover").mouseenter(function() {
					$(this).children('img').show();
				});
				$(".chess.hover").mouseleave(function() {
					$(this).children('img').hide();
				});
				$(".chess.hover").click(function() {
					var name = $(this).get(0).id;
					update(Number(name[5]), Number(name[6]));
				});
			}
			else {
				if (!(stones[0] == 0 || stones[1] == 0 || stones[0] + stones[1] == 64 || (validmoves[0] == 0 && validmoves[1] == 0)))
				{
					$.post('http://172.31.115.212:8888/', {"board": get_board(), "wait_time": 3}, 
					function(data, status){
						if (status == 'success'){
							console.log(data);
							move = get_move(data["move"]);
							update(move[0], move[1]);
						}
					});
				}
			}
		}
	}
}

function get_board() {
	var board="$$$$$$$$$$"
	for (var i = 0; i < 8; i++) {
		board += "$";
		for (var j = 0; j < 8; j++) {
			board += gameboard[7-i][j];
		}
		board += "$";
	}
	board += "$$$$$$$$$$ ";
	board += currentplayer;
	return board;
}

function get_move(move) {
	return [7 - Math.floor((move-11)/10), (move-11)%10];
}

/*
  Change the pieces in certain directions between target point and end point.
  The @current_piece is the current color the target point, or assumed color of it during idnetifying whether the point will a valid move.
  The @change is a boolean. True if it is a actual move, false if it is an assumed move.
*/
function change_inside(target_x, target_y, end_x, end_y, current_piece, change) {
	var i = Number(target_x);
	var j = Number(target_y);
	var listx = new Array();
	var listy = new Array();
	var num = Number(0);
	var found = Boolean(false);

	while (!found && (i != end_x || j != end_y)) {
		if (end_x - i != 0) {
			i = i + (end_x - i) / Math.abs(end_x - i);
		}
		if (end_y - j != 0) {
			j = j + (end_y - j) / Math.abs(end_y - j);
		}

		if (gameboard[i][j] == ".") {
			return false;
		} else if (gameboard[i][j] == current_piece) {
			found = true;
			break;
		}

		listx[num] = i;
		listy[num] = j;
		num++;
	}

	if (num == 0 || !found) {
		return false;
	} else if (change == true) {
		for (var n = 0; n < num; n++) {
			gameboard[listx[n]][listy[n]] = current_piece;
		}
	}
	return true;
}

/*
  Expand from the target point in 8 directions.
  The @current_piece is the current color the target point, or assumed color of it during idnetifying whether the point will a valid move.
  The @change is a boolean. True if it is a actual move, false if it is an assumed move.
*/
function expand(target_x, target_y, current_piece, change) {
	var enable = Boolean(false);
	var up = Number(Math.min(target_x, target_y));
	var down = Number(Math.min(7 - target_x, 7 - target_y));
	var left = Number(Math.min(target_x, 7 - target_y));
	var right = Number(Math.min(7 - target_x, target_y));

	if (change_inside(target_x, target_y, 0, target_y, current_piece, change))
		enable = true;
	if (change_inside(target_x, target_y, 7, target_y, current_piece, change))
		enable = true;
	if (change_inside(target_x, target_y, target_x, 0, current_piece, change))
		enable = true;
	if (change_inside(target_x, target_y, target_x, 7, current_piece, change))
		enable = true;
	if (change_inside(target_x, target_y, target_x - up, target_y - up, current_piece, change))
		enable = true;
	if (change_inside(target_x, target_y, target_x - left, target_y + left, current_piece, change))
		enable = true;
	if (change_inside(target_x, target_y, target_x + down, target_y + down, current_piece, change))
		enable = true;
	if (change_inside(target_x, target_y, target_x + right, target_y - right, current_piece, change))
		enable = true;

	if (enable == true) {
		return true;
	} else {
		return false;
	}

}

/*
  mark the valid possible moves on the xboard and oboard;
*/
function validateAndCount() {
	stones[0] = Number(0);
	stones[1] = Number(0);
	validmoves[0] = Number(0);
	validmoves[1] = Number(0);
	for (var i = 0; i < 8; i++) {
		for (var j = 0; j < 8; j++) {
			for (var k = 0; k < 2; k++) {
				validboard[k][i][j] = ".";
				if (gameboard[i][j] == ".") {
					if (expand(i, j, pieces[k], false)) {
						validboard[k][i][j] = pieces[k];
						validmoves[k]++;
					}
				} else if (gameboard[i][j] == pieces[k]) {
					stones[k]++;
				}
			}
		}
	}
}

function update(x, y) {
	gameboard[x][y] = currentplayer;
	//update the gameboard
	expand(x, y, currentplayer, true);
	// console.log((currentplayer == "x" ? 1 : 0) + " " + y + " " + (7-x) + " \r\n");
	record += (currentplayer == "x" ? 1 : 0) + " " + y + " " + (7-x) + " \r\n";

	//update the xboard, oboard, xmove and omove;
	validateAndCount();

	if (currentplayer == pieces[0]) {
		if (validmoves[1] != 0) {
			currentplayer = pieces[1];
		}
	} else if (currentplayer == pieces[1]) {
		if (validmoves[0] != 0) {
			currentplayer = pieces[0];
		}
	}

	display();
	handleEvent();

	checkWin();
}

function checkWin() {
	if (stones[0] == 0 || stones[1] == 0 || stones[0] + stones[1] == 64 || (validmoves[0] == 0 && validmoves[1] == 0)) {
		// TODO: add congratulations for the winner
		// Example:
		if (stones[0] > stones[1]) {
			setTimeout('alert("Black Wins!");',200);
		} else if (stones[0] < stones[1]) {
			setTimeout('alert("White Wins!");',200);
		} else {
			setTimeout('alert("Draw!");',200);
		}

		var d = new Date();
		dateStr = d.toISOString().substring(0,19).replace(/-|:/g, "");
		$('<a role="button" class="btn btn-success" id="downloadRecord" download="record_' + dateStr + '.txt" href="data:text/plain;charset=utf-8,' + encodeURIComponent(record) + '">Download Record</a>').insertAfter("#restart");
	}
}


$(document).ready(function() {

	user_is_black = true;

	gameboard = new Array(); // gameboard

	pieces = new Array("x", "o"); //pieces color: black="x" white="o" 黑子是叉，白子是圈
	currentplayer = pieces[0];
	stones = new Array(Number(0), Number(0)); //stone number for each player
	validmoves = new Array(Number(0), Number(0)); // possible moves for each player
	validboard = new Array(new Array(), new Array()); // the possible moves's board for each player

	elementName = new Array("black", "white");
	photoScource = new Array("./img/black.png", "./img/white.png", "./img/black2.png", "./img/white2.png");

	record = ""; // record string

	for (var i = 0; i < 8; i++) {
		gameboard[i] = new Array();
		validboard[0][i] = new Array();
		validboard[1][i] = new Array();

		for (var j = 0; j < 8; j++) {
			gameboard[i][j] = ".";
			validboard[0][i][j] = ".";
			validboard[1][i][j] = ".";
		}
	}

	oldboard = new Array();
	for (var i = 0; i < 8; i++) {
		oldboard[i] = new Array();
	}

	restart();
	$("#restart").attr('onclick', 'restart()');

});
